Automatic aspect ratio switch during FMV playback

FlatOutPS2 opened the pull request:
Add option to Emulation Settings - GS Window to for automatically switching to 4:3 aspect ratio when a FMV plays, and back to widescreen/stretch when it's finished.

![aspectswitch](https://cloud.githubusercontent.com/assets/15942346/14832352/1c40903e-0bfa-11e6-9067-34c7b08f2fe5.png)

Code probably requires optimising to follow PCSX2 coding guidelines, as I'm not a PCSX2 code or C++ expert.

Feature-request by trostboot as posted in issues: #1212

Update: latest commit Appveyor build link(April 29th): https://ci.appveyor.com/api/buildjobs/9lu8c562g7cuy52a/artifacts/pcsx2-v1.5.0-dev-678-gf6c11a1-master-933-vs2015-Win32-AppVeyor.7z

ssakash commented:
I'd prefer if this gets changed to a combobox, that way users could choose which aspect ratio to change into during FMV's. Any thoughts ?

On the other hand, I really think it'd be nice to properly align the dialog elements at GS window tab.

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -536,6 +538,27 @@ void DoFmvSwitch()
+	}
+
+	if (GSFrame* gsFrame = wxGetApp().GetGsFramePtr())
+		if (GSPanel* woot = gsFrame->GetViewport())
```
Please use something more informative than ``woot`` for the variable name. Maybe ``viewport`` ?

FlatOutPS2 replied:
That's code I copied from the elsewhere in PCSX2, it's actually used elsewhere in the same file too. That's why I left the name alone, though I do agree a more informative name would be better.

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -563,6 +586,23 @@ void Pcsx2App::LogicalVsync()
+			int diff = cpuRegs.cycle - eecount_on_last_vdec;
+			if (diff > 60000000) {
+				Console.Warning("FMV off");
+				Do43Switch(0);
```
This console warning might not be useful for the users,  Maybe change it to something so they it only triggers for dev builds and debug builds ? (In case you need it for debugging)

FlatOutPS2 replied:
That's code I copied from the "switch to software for FMV" gamefix, it can be removed as far as I'm concerned.

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -536,6 +538,27 @@ void DoFmvSwitch()
 	paused_core.AllowResume();
 }
 
+void Do43Switch(int on)
```
Why is on an ``int`` ? Better to change it to ``bool``.

FlatOutPS2 replied:
Good point. :p

FlatOutPS2 commented:
> I'd prefer if this gets changed to a combobox, that way users could choose which aspect ratio to change into during FMV's. Any thoughts ?

The current option seemed best to me because I didn't see why anyone would want to choose to go to any other aspect ratio during FMV's.

ssakash commented:
Current checkbox is fine if you think people wouldn't want to use other aspect ratios on FMV's. (I hope that's the case )

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -563,6 +586,27 @@ void Pcsx2App::LogicalVsync()
+#ifdef PCSX2_DEVBUILD
+				Console.Warning("FMV off");
+#endif
+				Do43Switch(0);
```
I prefer a ``DevCon.Warning`` over a ``#ifdef`` for dev build.

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -563,6 +586,27 @@ void Pcsx2App::LogicalVsync()
+#ifdef PCSX2_DEVBUILD 
+			Console.Warning("FMV on");
+#endif
+			Do43Switch(1);
```
Would be nice to use ``true`` for ``1`` and ``false`` for ``0`` on function call for ``Do43Switch ()``. It's a Bool parameter after all ;)

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -563,6 +586,27 @@ void Pcsx2App::LogicalVsync()
 	}
 
+	if (g_Conf->GSWindow.IsToggleAspectRatioSwitch) {
+		if (EnableFMV == 1) {
```
Just ``if (EnableFMV)`` is enough. The ``== 1`` might make people think it's an ``int`` whereas it's only a ``bool``. I suggest to use the bool types ``true`` or ``false`` whenever possible when dealing with Boolean variables

FlatOutPS2 replied:
Want me to update that for the "Switch to GSdx software" gamefix too? That's where I'm borrowing it from.

ssakash replied:
Yeah, better to also change that to make the code consistent.

ssakash commented on pcsx2/gui/Panels/GSWindowPanel.cpp:
```diff
@@ -49,6 +49,7 @@ Panels::GSWindowSettingsPanel::GSWindowSettingsPanel( wxWindow* parent )
 	m_check_Fullscreen	= new pxCheckBox( this, _("Default to fullscreen mode on open") );
 	m_check_VsyncEnable	= new pxCheckBox( this, _("Wait for Vsync on refresh") );
 	m_check_DclickFullscreen = new pxCheckBox( this, _("Double-click toggles fullscreen mode") );
+	m_check_AspectRatioSwitch = new pxCheckBox(this, _("Switch to 4:3 aspect ratio when a FMV plays"));
```
``When a FMV plays`` -> ``When an FMV plays``. Other than that,  code looks good to me 👍

willkuer commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -547,18 +570,35 @@ void Pcsx2App::LogicalVsync()
 			Console.Warning("FMV on");
 			DoFmvSwitch();
-			EnableFMV = 0;
+			EnableFMV = false;
```
This will make your toggling only work if you did not enable the switch to sw hack. I would suggest to merge the enabling and disabling of both hacks


Maybe something like that:
```
if (EmuConfig.Gamefixes.FMVinSoftwareHack || g_Conf->GSWindow.IsToggleAspectRatioSwitch) {
    if (EnableFMV) {
        Console.Warning("FMV on");
        if (EmuConfig.Gamefixes.FMVinSoftwareHack) 
            DoFmvSwitch();
        if (g_Conf->GSWindow.IsToggleAspectRatioSwitch)
            Do43Switch();
        EnableFMV = false;
    }
}
```

and the same for the disabling.

If it was intended to keep the aspect ratio on sw renderer ignore me.

willkuer commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -547,18 +570,35 @@ void Pcsx2App::LogicalVsync()
 
 		if (FMVstarted){
 			int diff = cpuRegs.cycle - eecount_on_last_vdec;
 			if (diff > 60000000 ) {
```
What kind of voodoo is that actually?
Is this an empirical number?
Would be interesting to know whether the sw-fmv switch would still lead to random crashing if this number would be different.

ssakash replied:
I was wondering the same actually... Maybe it's possible that all FMV's consume the specific cycle interval of the EE. Not sure though... I've never heard cases of random crashing though.

willkuer replied:
It sounds a lot like the ntsc base frequency multiplied by number of cpu ticks within a second. So I guess it means one frame after the fmv ends... but I don't know. Sounds awful.

willkuer commented:
I think this option should always be turned on, or? I guess all or almost all pre-rendered fmv's need 4:3? Accordingly I would change the default value to true and thus auto-enable it.

FlatOutPS2 commented:
> I think this option should always be turned on, or? I guess all or almost all pre-rendered fmv's need 4:3? Accordingly I would change the default value to true and thus auto-enable it.

I've updated to prevent accidentally triggering the option when 4:3 is selected(either in menu or with F6), so it supports auto-enabling. But I'm not sure if there are games with a 16:9 widescreen option that have 16:9 fmv's. If there are, it shouldn't be auto-enabled.

willkuer commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -542,11 +542,13 @@ void Pcsx2App::LogicalVsync()
 	if (EmuConfig.Gamefixes.FMVinSoftwareHack || g_Conf->GSWindow.IsToggleAspectRatioSwitch) {
+		AspectRatioType& art = g_Conf->GSWindow.AspectRatio;
+		if (g_Conf->GSWindow.IsToggleAspectRatioSwitch && !EnableFMV && !FMVstarted && art == AspectRatio_4_3 && prevAR != art)
+			prevAR = art;
```
```
If(prevAR != art)
    prevAR = art;
```
is the same as 
```
prevAR = art;
```

But do you really need this? It is only active while not within FMV. But after FMVenable=true it will automatically be overwritten in L553 or is set due to L76.

FlatOutPS2 replied:
It's needed for when you manually switch to 4:3 aspect ratio after using the widescreen option before. For example when switching games without shutting down PCSX2.

willkuer commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -542,11 +542,13 @@ void Pcsx2App::LogicalVsync()
-				AspectRatioType& art = g_Conf->GSWindow.AspectRatio;
 				if (art != AspectRatio_4_3) {
 					prevAR = art;
 					art = AspectRatio_4_3;
```
Also
```
if (art != AspectRatio_4_3) {
    prevAR = art;
    art = AspectRatio_4_3;
}
```
is the same as 
```
prevAR = art;
art=AspectRatio_4_3;
```
if `prevAR` is initialized as `AspectRatio_4_3`

FlatOutPS2 replied:
No, it only does that when art is different from prevAR. If the user has selected a different aspect ratio in the menu, it will no longer use the original initialization. It's to catch possible manual changes from the menu aspect ratio or switching with F6 to 4:3, preventing undesired changes from or to widescreen.

MrCK1 commented:
Is it necessary to account for widescreen patches in relation to auto-switching? For instance, going from an FMV to 3D?

FlatOutPS2 commented:
> Is it necessary to account for widescreen patches in relation to auto-switching? For instance, going from an FMV to 3D?

No. It's based on aspect ratio of the PCSX2 window, not the internal game aspect ratio. It will work for any game, regardless of patches or ingame widescreen options.

MrCK1 commented:
> No. It's based on aspect ratio of the PCSX2 window, not the internal game aspect ratio. It will work for any game, regardless of patches or ingame widescreen options.

OK, sounds good then.

gregory38 commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -546,19 +541,46 @@ void Pcsx2App::LogicalVsync()
-			Console.Warning("FMV on");
-			DoFmvSwitch();
-			EnableFMV = 0;
+	if (EmuConfig.Gamefixes.FMVinSoftwareHack || g_Conf->GSWindow.IsToggleAspectRatioSwitch) {
```
Please add back the DoFmvSwitch function. Add a on_off boolean parameter. This way code can factorized a little.

FlatOutPS2 commented:
> Please add back the DoFmvSwitch function. Add a on_off boolean parameter. This way code can factorized a little.

Is this to your liking?

danilaml commented on pcsx2/gui/AppConfig.cpp:
```diff
@@ -883,6 +884,7 @@ void AppConfig::GSWindowOptions::LoadSave( IniInterface& ini )
 	IniEntry( IsFullscreen );
 
     IniEntry( IsToggleFullscreenOnDoubleClick );
+	IniEntry(IsToggleAspectRatioSwitch);
```
mixed indentation in this file

FlatOutPS2 replied:
Looks fine in VS. 

It seems to be because the indentation for FullscreenOnDoubleClick was created with spaces, unlike the other other lines which use a tab.

danilaml replied:
You might as well fix it (and also, add spaces around inientry arg,to be more consistent with surrounding style).
pcsx2 coding style is something nonexistent.

FlatOutPS2 replied:
Yeah, I caught that too, I've updated it to add to a future commit

gregory38 commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -531,6 +531,42 @@ extern bool FMVstarted;
+				prevAR = art;
+				art = AspectRatio_4_3;
+			}
+			if (GSFrame* gsFrame = wxGetApp().GetGsFramePtr())
```
The part that you do in both case must be outside of the on_off if.

And you could use the else statement.
```
if (on_off){
} else {
}
```

turtleli commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -529,10 +531,34 @@ extern bool FMVstarted;
 {
-	ScopedCoreThreadPause paused_core( new SysExecEvent_SaveSinglePlugin(PluginId_GS) );
-	renderswitch = !renderswitch;
+	ScopedCoreThreadPause paused_core(new SysExecEvent_SaveSinglePlugin(PluginId_GS));
```
While it's necessary to pause and unpause for the "switch to software" hack, I don't think it's necessary if only the aspect ratio is changing.

FlatOutPS2 replied:
I had some issues with that when I started out with this PR, so I decided to keep it this way. But I just tested on the current code with it restricted to the software game fix and it works without issue now. I'll update.

turtleli commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -529,10 +531,34 @@ extern bool FMVstarted;
 extern bool EnableFMV;
 
-void DoFmvSwitch()
+void DoFmvSwitch(bool on_off)
```
Maybe a different variable name would be better i.e. `on`? `on_off` doesn't really make sense.

FlatOutPS2 replied:
I just copied how gregory38 described it :p , I'll just call it `on` again.

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -529,11 +531,34 @@ extern bool FMVstarted;
+		else {
+			if (art == AspectRatio_4_3 && prevAR != AspectRatio_4_3)
+				art = prevAR;
+		}
```
Maybe use an ``else if ()``instead ?

refractionpcsx2 commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -544,7 +544,7 @@ void DoFmvSwitch(bool on)
 			}
 		}
-		else {
+		else if (!on) {
```
isn't that moot? it's a bool, it's either on or it isn't

FlatOutPS2 replied:
ssakash said:
> Maybe use an else if ()instead ?

So...

ssakash replied:
I actually wanted you to make the following change:

**old**:
``else``
{
``if ()``
}

**New**:

``else if ()``
{
}

But you did this instead : 
``else if ()``
{
``if ()``
}

ssakash commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -529,11 +531,34 @@ extern bool FMVstarted;
+			if (art == AspectRatio_4_3 && prevAR != AspectRatio_4_3)
+				art = prevAR;
+		}
+
```
You don't need the check for ``!on`` , you  can replace it with the condition for inner ``if ()`` check.

FlatOutPS2 replied:
That wasn't exactly clear from what you said earlier.

ssakash replied:
Sorry , I was a bit lazy so I didn't explain it clearly. :P

gregory38 replied:
Anyway, that is a bad solution.
Keep
if (on) {
 do thing
} else {
 do other thing
}
It isn't a good idea to merge other thing into an else if because it is 2 different/independent actions (It is hard to explain).

elseif is better for
if (a == 0)
elseif (a == 1)
else

ssakash replied:
Fair enough , I guess. But I still don't think using ``else if ()`` at this scenario compromises readability.

turtleli commented:
Note: you need to rebase properly, there's unrelated stuff in your PR (I see stuff from my commits). See https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History (I'm sure someone can probably give you more specific instructions if you need it).

willkuer commented:
The most important thing here is that you need to force push (push --f) after you squashed/rebased to overwrite the commited history and avoid merging conflicts. If everything looks perfect in your local git you can force push.

FlatOutPS2 commented:
Not sure how it happened, but it should be good now.

turtleli commented on pcsx2/IPU/IPU.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "Gif.h"
 #include "Vif_Dma.h"
 #include <limits.h>
+#include "GSFrame.h"
```
I think the header file you were looking for is AppConfig.h

FlatOutPS2 replied:
Looks like it. It accepted GSFrame.h, so I just assumed it was correct. :)

gregory38 commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -529,11 +531,32 @@ extern bool FMVstarted;
+				prevAR = art;
+				art = AspectRatio_4_3;
+			}
+		}
```
To be consistent you need to put the else on the same line so 
```
if (test) {
;
} else {
;
}
```

gregory38 commented:
Just need to change the `else`. Otherwise code seem fine for me (didn't test it). I let the expert handle remaining reviewing/merging.

FlatOutPS2 commented:
Not sure if the` else `was my fault or VS's. VS puts it on the next line automatically. Anyway, it's been corrected.

turtleli commented:
Hmm. Doesn't work on Suikoden 4 (stays at 4:3 after I load a save game), but works on the other stuff I've tried (Suikoden 5, Suikoden Tactics, Growlanser, Xenosaga 2). That's probably due to how the FMV detection is done though.

FlatOutPS2 commented:
> That's probably due to how the FMV detection is done though.

Yeah, I remember having the same issue a while ago with the switch to software gamefix in one or two games.

turtleli commented:
Correction: It partially works on Suikoden 4, but it depends on what part of the game it is - I just happened to load the game to an area that it doesn't work with ;) (It also happens to be one of the saves I check for GSdx improvements/regressions) Accessing the menu does switch it out of 4:3 mode.

One issue though - if I close PCSX2 during an FMV, the next time I restart PCSX2 the aspect ratio will be set to 4:3.

ssakash commented:
In that case I guess the aspect ratio change should be temporary,  solution would be to not write the changes to the INI value and rework ``DoResize()`` with an extra parameter to allow temporary change of the aspect ratio ignoring the INI entry value.

FlatOutPS2 commented:
Rewrite complete, the Ini setting isn't touched anymore. And so far no regressions anymore either.

turtleli commented:
Your second commit suggests that your first commit has crashing issues. Is this because
`Resize -> Toggle HW/SW` causes problems that `Toggle HW/SW -> Resize` avoids?

turtleli commented on pcsx2/gui/FrameForGS.cpp:
```diff
@@ -155,11 +155,22 @@ void GSPanel::DoResize()
+	extern AspectRatioType iniAR;
+	extern bool switchAR;
 	double targetAr = clientAr;
-	if( g_Conf->GSWindow.AspectRatio == AspectRatio_4_3 )
```
I'm wondering whether you could just extend the if so it's
`if( g_Conf->GSWindow.AspectRatio == AspectRatio_4_3 || (g_Conf->GSWindow.IsToggleAspectRatioSwitch && FMVswitch))`

It may simplify your current solution (of course, I may have overlooked something too).

FlatOutPS2 replied:
FMVstarted always runs for as long as you run the FMV and affects both the FMV 43 switch and FMV software switch, switchAR is just for the FMV 43 switch, and can be turned off if you want to manually change the aspect ratio during a FMV.

FlatOutPS2 commented:
> Your second commit suggests that your first commit has crashing issues. Is this because
Resize -> Toggle HW/SW causes problems that Toggle HW/SW -> Resize avoids?

Yes, but also the other way around depending on if it's activating or deactivating the switches. And it only causes a crash when in full screen and you have the aspect ratio set to fit to screen. Otherwise it was fine.

turtleli commented:
Hmm. I'm not sure what to make of the F6 switching behaviour. If the 4:3 switch is on and you press F6 during an FMV - should it cycle through the different aspect ratios (current PR behaviour) or should it revert to the aspect ratio that the user normally uses first and then cycle on subsequent presses? I think the latter makes more sense, but I normally just use 4:3 and don't change aspect ratios.

> Yes, but also the other way around depending on if it's activating or deactivating the switches. And it only causes a crash when in full screen and you have the aspect ratio set to fit to screen. Otherwise it was fine.

Is there an explanation why this is the case? The toggle SW/HW hack is fragile, but having it crash/not crash depending on the order of the switch and resize and whether the hack is activating or deactivating doesn't make much sense to me. What OS/GPU/monitor resolution are you using? Did the crashing occur on specific games or just in general? Did you try using `viewport->PostSizeEvent()` instead of `viewport->DoResize()` so that the resize is queued up instead of done immediately (it may not make a difference)?

FlatOutPS2 commented:
> Hmm. I'm not sure what to make of the F6 switching behaviour. If the 4:3 switch is on and you press F6 during an FMV - should it cycle through the different aspect ratios (current PR behaviour) or should it revert to the aspect ratio that the user normally uses first and then cycle on subsequent presses? I think the latter makes more sense, but I normally just use 4:3 and don't change aspect ratios.

Your suggestion makes sense, but if you select fit to screen and want to go to 16:9 this does require more actions, 3(43>FTS>43 again>16:9) instead of a maximum of 2(43>16:9>FTS) with my implemention.

> Is there an explanation why this is the case? The toggle SW/HW hack is fragile, but having it crash/not crash depending on the order of the switch and resize and whether the hack is activating or deactivating doesn't make much sense to me. What OS/GPU/monitor resolution are you using? Did the crashing occur on specific games or just in general? Did you try using viewport->PostSizeEvent() instead of viewport->DoResize() so that the resize is queued up instead of done immediately (it may not make a difference)?

I have an AMD GPU, Windows 7 and a 16:10 monitor. It crashed on all games.
I tried using PostSizeEvent and it made it crash as well, though at the start where the DoResize crashed at the end of an FMV. Have you tried replicating the issue with the first commit yourself?

turtleli commented:
> Your suggestion makes sense, but if you select fit to screen and want to go to 16:9 this does require more actions, 3(43>FTS>43 again>16:9) instead of a maximum of 2(43>16:9>FTS) with my implemention.

Well, I assume that someone who normally uses fit to screen will probably prefer fit to screen instead of 16:9. Either way is fine though.

> I have an AMD GPU, Windows 7 and a 16:10 monitor. It crashed on all games.
> I tried using PostSizeEvent and it made it crash as well, though at the start where the DoResize crashed at the end of an FMV. Have you tried replicating the issue with the first commit yourself?

Windows 10, NVIDIA, 16:10. I only get issues that I get on master too, your first commit doesn't cause issues either. That it only crashes when you use fit to screen makes me wonder whether it's partially driver induced (like #1075).

It probably would be nice to have other testers (I'll do a Linux test later) comment on whether they have issues or not.

turtleli commented:
I tested both commits on Linux and I didn't have issues either. I suppose it's fine since the option on it's own isn't problematic.

I'll have this merged in a few days if no one complains of any issues.

FlatOutPS2 commented:
It seems my commit wasn't causing the crash. It happens with just the software switch game fix as well, even on the stable v1.4.0. It seems to only happen with OpenGL HW renderer and fit to screen aspect ratio, and more often at higher than native resolution after switching from windowed to fullscreen mode.

For some reason my fix helps when both the 4:3 switch and software switch are on, but if it's just the switch to software gamefix, it still crashes... :S

refractionpcsx2 commented:
Yeh there's an issue with switching from software to hardware in opengl on some games... I can't remember what ones though, but I guess you have one lol

FlatOutPS2 commented:
> Yeh there's an issue with switching from software to hardware in opengl on some games... I can't remember what ones though, but I guess you have one lol

Some games? I happened to all the games I tried. :p

turtleli commented:
All your games are broken :P

It's probably a system dependant thing - Maybe run in debug mode and look at the stack traces when it crashes to see if it provides any clues?

refractionpcsx2 commented:
I just tried one game at random and it didn't crash :P

can you list some of the games it definitely happens on and your exact GS settings so we can try to recreate?

FlatOutPS2 commented:
1. Turn on switch to software for FMV switch in game fixes
2. Set Renderer to OpenGL (Hardware).
3. Set Internal Resolution to anything above native, leave everything else to default(different settings don't seem to matter though).
4. Set aspect ratio to Fit to Window/Screen
5. Start a game with FMV's.
6. Check if software switch works for the FMV you're testing it on.
7. Go fullscreen, start a movie and if it doesn't crash immediately, stop the FMV
8. If it doesn't crash, try step 7 a few more times.
9. If it still doesn't crash, go back to window mode and back to fullscreen and try again.

It crashes on everygame for me. Just tried Burnout 3, crashed on the EA logo...

refractionpcsx2 commented:
> Go fullscreen, start a FMV and if it doesn't crash immediately, stop the FMV
> If it doesn't crash, try step 7 a few more times.

So it doesn't crash all the time on all games like you stated then? It only crashes if you f### around with it a load? :P

FlatOutPS2 commented:
> So it doesn't crash all the time on all games like you stated then? It only crashes if you f### around with it a load?

No, it often crashes on first time, just occasionally not.

refractionpcsx2 commented:
I just followed your steps on burnout 3 as you said it works on there, had the switch on, scailing to x2, blending and depth options all on, went in to full screen, back to windowed loads and skipped FMV's, nothing I could do would make it crash.

gregory38 commented:
AMD driver powa !

@FlatOutPS2 
Could you try to add 
```
glFinish();
```
at the end of GSDeviceOGL desctructor (around line 136). Let's say after `m_shader = NULL;`

glFinish will flush the rendering gl command otherwise the driver could still be busy. I'm nearly sure it won't have any impact but worth a try.

FlatOutPS2 commented:
> I'm nearly sure it won't have any impact but worth a try.

You were right, no impiact. It still crashes. :p

refractionpcsx2 commented:
Don't suppose you can run it  under the VS local debugger can you? :P

FlatOutPS2 commented:
Stack traces don't mention GSdx when it crashes.
It crashes after it does the ScopedCoreThreadPause paused_core in DoFMVswitch
It ends up stalling in:

w32_CancelableWait.c

line 88:		  status = WaitForMultipleObjects (nHandles, handles, PTW32_FALSE, timeout);

		PTW32_FALSE	PTW32_FALSE (0)	<unnamed-enum-PTW32_FALSE>
+		handles	0x0017ecd8 {0x000002d4, 0xffffffff}	void *[2]
		handles[1]	0xffffffff	void *
		nHandles	1567776	unsigned long

turtleli commented:
Do any of the other threads have useful info? The Threads window should show you the stacktraces for all the threads.

FlatOutPS2 commented:
I didn't see anything special in the other threads. Might just be me though. :p

If I set viewport.x (or y) in DoResive(FrameForGS.cpp)to anything that makes it not exactly fullscreen, it doesn't crash anymore. I changed 1920 to 1921 and no crashes.

turtleli commented:
Could you add a comment to the code about the fit to screen crashing and the workaround (making the GSPanel not cover the full screen when switching from HW to SW and vice versa)? I suspect you would also have issues with 16:9 if you were using a 16:9 resolution too.

FlatOutPS2 commented:
> Could you add a comment to the code about the fit to screen crashing and the workaround (making the GSPanel not cover the full screen when switching from HW to SW and vice versa)? I suspect you would also have issues with 16:9 if you were using a 16:9 resolution too.

I set the screen resolution of Windows to a 16:9 aspect ratio and it did crash at PCSX2's 16:9 resolution.
I removed the original crash fix and replaced it by a possible workaround. It sets the zoom to 101/99% if the switch to software fix is on in fullscreen and the aspect ratio fits the screen.

turtleli commented on pcsx2/gui/FrameForGS.cpp:
```diff
@@ -171,6 +182,13 @@ void GSPanel::DoResize()
+		if (zoom < 1.001)
+			zoom = (g_Conf->GSWindow.Zoom.ToFloat() + 1.0) / 100.0;
+		if (zoom > 0.999)
+			zoom = (g_Conf->GSWindow.Zoom.ToFloat() - 1.0) / 100.0;
```
Shouldn't it be
```
if (zoom > 0.999 && zoom < 1.001)
	zoom = (g_Conf->GSWindow.Zoom.ToFloat() - 1.0) / 100.0;
```
? Otherwise if `g_Conf->GSWindow.Zoom.ToFloat()` is 101 you'll end up with a zoom of 1.

FlatOutPS2 replied:
I was actually going for  +1 for zoom values for above and equal to 1 and below 1.001, and -1 for zoom values below 1 and above 0.999, but after I updated the fix I forgot to add the catch for the 1 value back to the if statements.

turtleli commented on pcsx2/gui/FrameForGS.cpp:
```diff
@@ -171,6 +182,13 @@ void GSPanel::DoResize()
 		zoom = std::max( (float)arr, (float)(1.0/arr) );
+	if (arr == 1 && EmuConfig.Gamefixes.FMVinSoftwareHack && g_Conf->GSWindow.IsFullscreen) {
+		//avoids crash on some systems switching HW><SW in fullscreen aspect ratio's with FMV Software switch.
+		if (zoom < 1.001 && zoom >= 1.0000)
```
Could you switch the order to `zoom >= 1.0000 && zoom < 1.001` ? It's slightly easier to read that way.

And could you squash the 3rd commit into your 2nd commit? I can merge after that's done.

turtleli commented on pcsx2/gui/FrameForGS.cpp:
```diff
@@ -173,6 +184,8 @@ void GSPanel::DoResize()
 		zoom = std::max( (float)arr, (float)(1.0/arr) );
 
 	viewport.Scale(zoom, zoom*g_Conf->GSWindow.StretchY.ToFloat()/100.0 );
+	if (viewport.x == (double)client.GetWidth() && viewport.y == (double)client.GetHeight() && EmuConfig.Gamefixes.FMVinSoftwareHack && g_Conf->GSWindow.IsFullscreen)
```
`viewport == client`?

FlatOutPS2 replied:
Viewport values can get updated at lines 178, 180 and 186, the if statement checks if it's still the same as client after those possible updates. If it is, it increases the size of the viewport to not exactly equal client size, thus preventing the crash.

turtleli replied:
Sorry, I wasn't clear. You could replace
`viewport.x == (double)client.GetWidth() && viewport.y == (double)client.GetHeight()`
with
`viewport == client`

FlatOutPS2 replied:
Oh, right. Of course. :D

Has been updated.
